frame/dvm/src/lib.rs

```diff
@@ -401,49 +469,74 @@ impl<T: Config> Pallet<T> {
 			}
 		}
 
-		if transaction.gas_limit >= T::BlockGasLimit::get() {
+		if gas_limit >= T::BlockGasLimit::get() {
 			return Err(InvalidTransaction::Custom(
 				TransactionValidationError::InvalidGasLimit as u8,
 			)
 			.into());
 		}
 
-		let account_data = <T as darwinia_evm::Config>::RingAccountBasic::account_basic(&origin);
-		let fee = transaction.gas_price.saturating_mul(transaction.gas_limit);
-		let total_payment = transaction.value.saturating_add(fee);
-		if account_data.balance < total_payment {
+		let base_fee = T::FeeCalculator::min_gas_price();
+		let mut priority = 0;
+		let gas_price = if let Some(gas_price) = transaction_data.gas_price {
+			// Legacy and EIP-2930 transactions.
+			// Handle priority here. On legacy transaction everything in gas_price except
+			// the current base_fee is considered a tip to the miner and thus the priority.
+			priority = gas_price.saturating_sub(base_fee).unique_saturated_into();
+			gas_price
+		} else if let Some(max_fee_per_gas) = transaction_data.max_fee_per_gas {
+			// EIP-1559 transactions.
+			max_fee_per_gas
+		} else {
 			return Err(InvalidTransaction::Payment.into());
-		}
+		};
 
-		let min_gas_price = T::FeeCalculator::min_gas_price();
-		if transaction.gas_price < min_gas_price {
+		if gas_price < base_fee {
 			return Err(InvalidTransaction::Payment.into());
 		}
 
-		Ok(account_data.nonce)
+		let mut fee = gas_price.saturating_mul(gas_limit);
+		if let Some(max_priority_fee_per_gas) = transaction_data.max_priority_fee_per_gas {
+			// EIP-1559 transaction priority is determined by `max_priority_fee_per_gas`.
+			// If the transaction do not include this optional parameter, priority is now considered zero.
+			priority = max_priority_fee_per_gas.unique_saturated_into();
+			// Add the priority tip to the payable fee.
+			fee = fee.saturating_add(max_priority_fee_per_gas.saturating_mul(gas_limit));
```

Comment:
I think
before this line we get
```
fee = gas_price * gas_limit
gas_price = base_fee_price + priority_fee_per_gas
```
so why here we need add max_priority_fee_per_gas.saturating_mul(gas_limit) again?

Reply:
Good question
1. For Legacy and EIP-2930 transactions, the `max_priority_fee_per_gas` field is None and `gas_price` is Some value.
2. For EIP-1559 Transaction, the `max_priority_fee_per_gas` field is Some, and `gas_price` is None.